_labs/lab08.markdown

```diff
@@ -0,0 +1,171 @@
+---
+layout: lab
+_id: "8"
+title: "Constraint-Based Analysis"
+synopsis: Writing a constraint-based static analysis for C programs with LLVM and Z3.
+---
+
+### Objective 
+
+In this lab, you will implement a constraint-based analysis to detect exploitable divide-by-zero bugs.
+A bug is exploitable if hackers can control inputs or environments, thereby triggering unintended behaviors (e.g., denial-of-service) through the bug.
+For example, [a recently reported divide-by-zero bug][bug] in the Linux kernel can be exploitable and crash the system.
+You will design a static analysis that detects such bugs by combining reaching definition analysis and taint analysis on top of a constraint solver, Z3.
+
+### Setup
+
+The skeleton code for Lab8 is located under `/cis547vm/lab8/`.
+We will frequently refer to the top level directory for Lab 8 as `lab8` when describing file locations for the lab. Open the `lab8` directory in VSCode following the Instructions from [Course VM document][course-vm-doc]
+
+The following commands set up the lab, using the [Cmake][Cmake ref]/[Makefile][Make ref] pattern seen before.
+
+```sh
+/lab8$ mkdir build && cd build
+/lab8/build$ cmake ..
+/lab8/build$ make
+```
+
+The above command will generate an executable file 'constraint' that checks whether the input program has an exploitable divide-by-zero bug:
+
+```sh
+/lab8$ cd ./test
+/lab8/test$ clang -emit-llvm -S -fno-discard-value-names -c simple0.c
+/lab8/test$ ../build/constraint simple0.ll
+```
+
+If you’ve done everything correctly up to this point you should see 'Potential divide-by-zero points:'. 
+
+### Lab Instructions
+
+In this lab, you will design a reaching definition analysis and taint analysis using [Z3][Z3Guide].
+The main tasks are to design the analysis in the form of Datalog rules through the [Z3 C++ API][Z3C++API], and implement a function that extracts logical constraints in the form of Datalog facts for each LLVM instruction.
+
+We will then feed these constraints, along with your Datalog rules, into the Z3 solver which should report any *exploitable* divide-by-zero errors.
```

Comment:
Edit:

We will use the datalog rules with the facts you extracted from each instruction to find any *exploitable* ...